core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +157,949 @@ impl Decode for BabeSeal {
+							"slot" => slot_num
+						);
+						return
+					}
+
+					let (header, body) = b.deconstruct();
+					let header_num = header.number().clone();
+					let pre_hash = header.hash();
+					let parent_hash = header.parent_hash().clone();
+
+					// sign the pre-sealed hash of the block and then
+					// add it to a digest item.
+					let to_sign = (slot_num, pre_hash, proof.to_bytes()).encode();
+					let signature = pair.sign(&to_sign[..]);
+					let item = <DigestItemFor<B> as CompatibleDigestItem>::babe_seal(BabeSeal {
+						proof,
+						signature: LocalizedSignature {
+							signature,
+							signer: pair.public(),
+						},
+						slot_num,
+						vrf_output,
+					});
+
+					let import_block: ImportBlock<B> = ImportBlock {
+						origin: BlockOrigin::Own,
+						header,
+						justification: None,
+						post_digests: vec![item],
+						body: Some(body),
+						finalized: false,
+						auxiliary: Vec::new(),
+						fork_choice: ForkChoiceStrategy::LongestChain,
+					};
+
+					info!(target: "babe",
+						  "Pre-sealed block for proposal at {}. Hash now {:?}, previously {:?}.",
+						  header_num,
+						  import_block.post_header().hash(),
+						  pre_hash
+					);
+					telemetry!(CONSENSUS_INFO; "babe.pre_sealed_block";
+						"header_num" => ?header_num,
+						"hash_now" => ?import_block.post_header().hash(),
+						"hash_previously" => ?pre_hash
+					);
+
+					if let Err(e) = block_import.import_block(import_block, Default::default()) {
+						warn!(target: "babe", "Error with block built on {:?}: {:?}",
+							  parent_hash, e);
+						telemetry!(CONSENSUS_WARN; "babe.err_with_block_built_on";
+							"hash" => ?parent_hash, "err" => ?e
+						);
+					}
+				})
+				.map_err(|e| {
+					warn!("Client import failed: {:?}", e);
+					consensus_common::ErrorKind::ClientImport(format!("{:?}", e)).into()
+				})
+		)
+	}
+}
+
+/// check a header has been signed by the right key. If the slot is too far in
+/// the future, an error will be returned. If successful, returns the pre-header
+/// and the digest item containing the seal.
+///
+/// This digest item will always return `Some` when used with `as_babe_seal`.
+//
+// FIXME #1018 needs misbehavior types
+#[forbid(warnings)]
+fn check_header<B: Block + Sized>(
+	slot_now: u64,
+	mut header: B::Header,
+	hash: B::Hash,
+	authorities: &[Public],
+	threshold: u64,
+) -> Result<CheckedHeader<B::Header, DigestItemFor<B>>, String>
+	where DigestItemFor<B>: CompatibleDigestItem,
+{
+	trace!(target: "babe", "Checking header");
+	let digest_item = match header.digest_mut().pop() {
+		Some(x) => x,
+		None => return Err(format!("Header {:?} is unsealed", hash)),
+	};
+
+	let BabeSeal {
+		slot_num,
+		signature: LocalizedSignature {signer, signature },
+		proof,
+		vrf_output,
+	} = digest_item.as_babe_seal().ok_or_else(|| {
+		error!(target: "babe", "Header {:?} is unsealed", hash);
+		format!("Header {:?} is unsealed", hash)
+	})?;
+
+	if slot_num > slot_now {
```

Comment:
i'm interested to see how the rel. time stuff shakes out here (beyond the scope of this PR)